### DISCUSSION TITLE
Client streaming from the web browser

### DISCUSSION THREAD

**Discussion by paralin** [connectrpc/connect-go#254] [Q&A] [answered]
Looking into "connect" currently - it's very nice, with clean generated code.

Just one question regarding client-side streams:

This page: https://connect.build/docs/go/streaming - says that:

> All that said, connect-go fully supports all three types of streaming. All streaming subtypes work with the gRPC, gRPC-Web, and Connect protocols.

This implies that `grpc-web` can be used to run client-stream RPCs with Connect, but according to the grpc-web docs:

https://github.com/grpc/grpc-web/blob/master/doc/streaming-roadmap.md

> Once the whatwg spec is finalized and upload-streaming is supported by major browsers (2020), we will add client-streaming support in gRPC-Web (and Envoy) which will also enable half-duplex bidi streaming support.

This indicates that **client-streaming is not implemented in grpc-web yet.**

So I am having trouble understanding how connect implements the client-streaming with grpc-web. Is it actually not implemented?

If not, that above sentence in the docs may be a bit confusing to others looking for client-streaming implementations (like me).

**Comment by akshayjshah** [comment, marked as answer]
Great question :) To me, that document is confusing because it uses "gRPC-Web" to refer to two different things: a protocol and a specific Javascript implementation.

The gRPC-Web _protocol_ supports unary, client streaming, and server streaming RPCs with HTTP/1.1 or HTTP/2. If you have an HTTP/2 connection, it also supports bidirectional streaming.

The `connect-go` _implementation_ uses `net/http`, which has all the capabilities needed for all four RPC types. So `connect-go` servers and clients will happily handle any RPC type using the gRPC-Web protocol.

The `grpc/grpc-web` _implementation_ runs in web browsers, which buffer request bodies and therefore can't do client or bidirectional streaming. The upcoming `connect-web` _implementation_ will have the same limitation.

(Also, I'm going to convert this issue to a discussion. I haven't used discussions before, but we're giving them a try in this repo and this is a really good question!)

**Reply by paralin** [reply]
Thanks for the detailed answer. The discussion format is definitely more applicable to this.

**Reply by akshayjshah** [reply]
Ok, awesome - glad I could help!
